doc/decisions/0_drafts/errors_core.md

```diff
@@ -0,0 +1,292 @@
+  Often this case can be excluded by static analysis, i.e., because of the preceding code it is impossible that the pointer is `NULL`.
+
+  ```c
+  int foo (KeySet * ks, Key * key) {
+    if (key == NULL) {
+      return -1;
+    }
+
+    // [...] could lots of code, as long as the Key * key doesn't change
+
+    const Key * found1 = ksLookupByName (ks, "/foo", 0);
+    if (found1 != NULL) {
+      // no need to worry about key == NULL, because we checked earlier
+      // also no need to worry about ks == NULL, because the ksLookupByName would have returned NULL
+      const Key * found2 = ksLookup (ks, key, 0);
+    }
+  }
+  ```
+
+- It is not always necessary to report separate error codes, for _every_ error case.
+  In many cases, the caller can distinguish after the fact, e.g., `keySetName` fails when the given name is invalid and when the given `Key` has a read-only name.
+  This can be distinguished after the fact, by checking whether the `Key` has a read-only name with `keyIsLocked`.
+
+- Callers normally don't care about the exact error, unless they can do something about it.
+  This could be:
+
+  - reporting the error to a user
+  - switching to a different code path
+  - (during debugging) changing the code
+
+- Functions can be written in a such a way that they detect and report all errors without permanent effects.
+
+- The functions that cannot use a `Key * errorKey`, are side-effect-free.
+  Therefore, the only possible cause of errors are invalid arguments.
+
+## Considered Alternatives
+
+### Extra argument
+
+All functions that report errors take a `int * error` argument to which errors can be written.
+
+This makes the API much more convoluted to use.
+
+A benefit would be the return value remains fully usable.
+
+### Thread Local Storage
+
+Use a field in thread local storage (TLS) to store errors similar to `errno`.
+
+This leaves the API design entirely open.
+
+TLS was only standardized in C11, while Elektra currently only requires C99.
+
+This has all the same issues as `errno`: needs to be cleared to detect errors, can only store one error (per thread), etc.
+
+### Redesign API
+
+Redesign the API, so that every API function only has maximal one error case next to memory errors.
+
+- `nameNew(name) -> name_object` only has the error case of invalid name
+- `keyNew(name_object)` won't have error cases anymore
+- `keySetName(name_object)` only has the error case of read-only names
+
+This can be very limiting and in some cases makes it very hard to find out what the exact problem was.
+Reducing a function to a single error value, makes sense if the errors are easy to distinguish.
+But for more complex preconditions, this can be annoying and inefficient to use.
+While the above example can be solved, by simple checking with `keyIsLocked`, there could less straightforward cases.
+For example, a function `foo` that takes two keynames, both of which must be valid.
```

Comment:
Please use examples from our API.

Reply:
I would, but there is currently no such function that I know of. Definitely no function takes two keynames and I can't think of one that takes two other arguments with complex preconditions.

I still wanted to give the example, because it shows that this is a valid option, even if we add a function with such complex preconditions in the future.